### PR TITLE
allow STOP_SENDING on implicitly opened streams

### LINKED ISSUE #2345: STOP_SENDING on implicitly opened streams is error-prone
> A stream ID that is used out of order results in all streams of that type with lower-numbered stream IDs also being opened.

> Receiving a STOP_SENDING frame for a locally-initiated stream that is “Ready” or not yet created MUST be treated as a connection error of type STREAM_STATE_ERROR

Streams can easily become open while still in the `Ready` state on the sender if the sender does not necessarily transmit data on streams in exactly the order in which they are allocated. If the receiving application decides to abandon such a stream and sends STOP_SENDING, a connection error results.

If STOP_SENDING is additionally permitted for streams in the `Ready` state which have been implicitly opened by some higher-numbered stream of the same type entering any other state, then implicitly-opened streams can be safely discarded without waiting for a STREAM or RESET_STREAM frame which the application is no longer interested in.

### PR DESCRIPTION
**Pull request by marten-seemann**
As suggested in https://github.com/quicwg/base-drafts/issues/2345#issuecomment-455448227.
Fixes #2345.

### CODE REVIEW

**Comment by mikkelfj** [comment]
I am not necessarily objecting to this, but before just merging this in, please review past discussions on this - it is not as trivial as it might seem.

https://github.com/quicwg/base-drafts/issues/1050
https://github.com/quicwg/base-drafts/issues/1797

**Comment by mikkelfj** [comment]
> What concrete problem do you see, @mikkelfj?

I don't see a concrete problem, but I do remember a lot of pitfalls, so I just wanted to make a note of it before anything gets merged. I have not yet read through the issues in detail.

One concern is race conditions but it might be captured by the proposed text, I'm just not sure yet.

Also, it is difficult to grasp exactly what is the difference between not yet created, created, Ready, peer has seen data from higher stream but ID is below MAX_STREAM_ID.

**Comment by mikkelfj** [comment]
propose:

Receipt of a STOP_SENDING frame is invalid for a locally-initiated stream that
has not yet been created or is in the "Ready" state (see
{{stream-send-states}}). Receiving a STOP_SENDING frame for a locally-initiated
stream that has not transitioned past Ready state MUST be treated as a connection error of type
STREAM_STATE_ERROR unless there is a higher numbered stream that has made this transition already.  A peer must have seen data on the same stream or a higher numbered stream of same type before sending STOP_SENDING. An endpoint that receives a STOP_SENDING frame for a
receive-only stream MUST terminate the connection with error STREAM_STATE_ERROR.

See also discussion on associated issue. Removing Ready is not safe as this can lead to speculative STOP_SENDING, for example based on assumed out of band knowledge of the senders local state.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -4491,10 +4491,9 @@ to abruptly terminate transmission on a stream.
 Receipt of a STOP_SENDING frame is invalid for a locally-initiated stream that
 has not yet been created or is in the "Ready" state (see
 {{stream-send-states}}). Receiving a STOP_SENDING frame for a locally-initiated
-stream that is "Ready" or not yet created MUST be treated as a connection error
-of type STREAM_STATE_ERROR.  An endpoint that receives a STOP_SENDING frame for
-a receive-only stream MUST terminate the connection with error
-STREAM_STATE_ERROR.
+stream that is not yet created MUST be treated as a connection error of type
```
Comment (kazuho) [inline draft-ietf-quic-transport.md]: I think the other approach we might consider is to state that a stream enters a Ready state when the endpoints send a frame containing a higher stream ID.

Taking that approach might help resolve @mikkelfj's concern that the definition is becoming loose.

**Comment by marten-seemann** [comment]
I don't think we gain anything by worrying about out-of-band knowledge about stream states, and I don't think that @mikkelfj's text suggestion is readable. The same would apply to MAX_STREAM_DATA frames for remote-opened streams as well.

**Comment by mikkelfj** [comment]
Note that just edited my text. I think @kazuho has a point. But it still needs to be clear what the remote endpoint must observe before sending.

**Comment by mikkelfj** [comment]
> The same would apply to MAX_STREAM_DATA 

I think that needs to be fixed as well. The frame might get lost due to race conditions.

**Comment by marten-seemann** [comment]
> > The same would apply to MAX_STREAM_DATA
>
> I think that needs to be fixed as well. The frame might get lost due to race conditions.

Can you elaborate where you see a race condition here?

**Comment by marten-seemann** [comment]
I don't see any race condition we need to fix here. If it was permissible to send MAX_STREAM_DATA x+4, no amount of reordering will make this a protocol violation. If it was not permissible, and the remote peer just happened to open x+4 while the MAX_STREAM_DATA frame was in flight, then the peer got lucky this one time and the protocol violation won't be detected. I don't think this is something we need to address though.

**Comment by mikkelfj** [comment]
>  If it was permissible to send MAX_STREAM_DATA x+4, no amount of reordering will make this a protocol violation

then you have to be able track infinite amount state because the peer controls MAX_STREAM_ID.

> If it was not permissible, and the remote peer just happened to open x+4 while the MAX_STREAM_DATA frame was in flight, then the peer got lucky this one time and the protocol violation won't be detected.

You can always get lucky, by it is sufficient to risk enforcement, and more importantly, it protects against tracking infinite state.

**Comment by marten-seemann** [comment]
> then you have to be able track infinite amount state because the peer controls MAX_STREAM_ID.

No, you don't, because you're allowed to silently drop MAX_STREAM_ID frames for closed streams. We have text about that already.

**Comment by mikkelfj** [comment]
> No, you don't, because you're allowed to silently drop MAX_STREAM_ID frames for closed streams. We have text about that already.

I am talking about not yet opened streams.
But is you said in the #2353 there is text protecting against this - only I believe that the text is not very precise.
Wrt. STOP_SENDING I basically want the same protection, and for both frames, I'd like it to be clear what the sender of these reverse frames need to observe before concluding that it is safe to send them.

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (Ralith) [reply draft-ietf-quic-transport.md]: Enters the Send state, I think you mean? Ready is the state that currently forbids STOP_SENDING.

That seems like an effective solution, although I find this use of "created" intuitive and consistent with the "Within each type, streams are created with numerically increasing stream IDs" language.

**Comment by Ralith** [comment]
I agree that MAX_STREAM_DATA and STOP_SENDING should have exactly the same rules for when they're accepted, and that MAX_STREAM_DATA's current language, while somewhat vague, permits the case that this issue enables for STOP_SENDING. When we settle on a way to phrase this, it might be nice to use exactly the same language for both frames to make that clear.

**Comment by martinthomson** [comment]
#2354 supercedes this.

**Pull request closed without merging**